### ISSUE TITLE
Asset Management Proposal

### ISSUE THREAD

**Issue by gblakeman** [jonathanawesome/blanket#7]
Below are some proposed updates to how the project’s local theme assets (not `uploads`) are managed to improve cache-busting and ensure CDN’s have the option of telling a browser to never bother checking assets for an update.

------------------

Ensure Wordpress is only using `dist` for theme images/assets (update internal helpers/prefs to use it instead of `_/img`). This change allows us to only worry about serving assets from a single directory inside the theme—helpful when configuring a CDN for static assets.

Enable filename hashing in webpack: `main.css --> main.233e113732703bee76a8.css`. Use the `manifest.json` inside Wordpress to provide access to the hashed filenames: `getHashedAsset` and `getHashedAssetWithPath` (new proposed helpers – see [my fork](https://github.com/gblakeman/blanket/blob/master/app/helpers.php)).

With this setup, Wordpress will always serve assets with cache-busting, unique filenames. The names only change when the contents of the file changes.

```html
<link rel='stylesheet' id='styles-css' href='http://blanket.local/wp-content/themes/blanket/dist/main.233e113732703bee76a8.css' type='text/css' media='all' />
…
<img src="http://blanket.local/wp-content/themes/blanket/dist/xo.2e24f23aadf0cb534f86046473c9b278.jpg" />
…
<script type='text/javascript' src='http://blanket.local/wp-content/themes/blanket/dist/main.e17e6e6093d1ba3c87f7.js'></script>
```

Lastly, add some on-the-fly image optimization through webpack using the `ImageminPlugin` and `imageminMozjpeg`.

**Comment by gblakeman** [comment]
I have this setup running on my fork and it’s working well. Ultimately I like the simplicity of only serving assets from a single directory in the theme and _knowing_ they have unique filenames, but I can appreciate if the increased complexity of the initial setup doesn’t feel warranted.

See PRs [#5](https://github.com/gblakeman/blanket/pull/5) and [#7](https://github.com/gblakeman/blanket/pull/7) for the bulk of the required changes (the restructuring of the file directories is just my opinionated taste and wouldn’t need to happen).

**Comment by jonathanawesome** [comment]
Yeah, this is fantastic. Cleans up the constants in functions...awesome. 

I'm a huge fan of tweaking SVGs manually, for various reasons. How do you find SVGO use in production?

I've been hanging onto the _ directory for years, for no reason other than it felt comfortable.

**Comment by gblakeman** [comment]
> Yeah, this is fantastic. Cleans up the constants in functions...awesome.

Sweet, I’ll get a version working on your repo.

> I'm a huge fan of tweaking SVGs manually, for various reasons. How do you find SVGO use in production?

We’ll find out 😂. I tend to tweak the hell out of them manually too, so this is an experiment for me.

> I've been hanging onto the _ directory for years, for no reason other than it felt comfortable.

Yeah, the directory stuff I did is all just mental comfort. I always find my projects gravitate toward the original `app`/Rails setup I started with years ago for no solid reason.
